packages/adblocker/src/engine/bucket/html.ts

```diff
@@ -0,0 +1,211 @@
+      deserialize: NetworkFilter.deserialize,
+      filters: [],
+      optimize: config.enableOptimizations ? optimizeNetwork : noopOptimizeNetwork,
+    });
+
+    this.cosmeticIndex = new ReverseIndex({
+      config,
+      deserialize: CosmeticFilter.deserialize,
+      filters: [],
+      optimize: noopOptimizeCosmetic,
+    });
+
+    this.unhideIndex = new ReverseIndex({
+      config,
+      deserialize: CosmeticFilter.deserialize,
+      filters: [],
+      optimize: noopOptimizeCosmetic,
+    });
+
+    if (filters.length !== 0) {
+      this.update(filters, undefined);
+    }
+  }
+
+  public update(
+    newFilters: (NetworkFilter | CosmeticFilter)[],
+    removedFilters: Set<number> | undefined,
+  ): void {
+    const networkFilters: NetworkFilter[] = [];
+    const exceptionFilters: NetworkFilter[] = [];
+    const cosmeticFilters: CosmeticFilter[] = [];
+    const unhideFilters: CosmeticFilter[] = [];
+
+    for (const filter of newFilters) {
+      if (filter.isNetworkFilter()) {
+        if (filter.isException()) {
+          exceptionFilters.push(filter);
+        } else {
+          networkFilters.push(filter);
+        }
+      } else if (filter.isCosmeticFilter()) {
+        if (filter.isUnhide()) {
+          unhideFilters.push(filter);
+        } else {
+          cosmeticFilters.push(filter);
+        }
+      }
+    }
+
+    this.networkIndex.update(networkFilters, removedFilters);
+    this.exceptionsIndex.update(exceptionFilters, removedFilters);
+    this.cosmeticIndex.update(cosmeticFilters, removedFilters);
+    this.unhideIndex.update(unhideFilters, removedFilters);
+  }
+
+  public serialize(buffer: StaticDataView): void {
+    this.networkIndex.serialize(buffer);
+    this.exceptionsIndex.serialize(buffer);
+    this.cosmeticIndex.serialize(buffer);
+    this.unhideIndex.serialize(buffer);
+  }
+
+  public getSerializedSize(): number {
+    return (
+      this.networkIndex.getSerializedSize() +
+      this.exceptionsIndex.getSerializedSize() +
+      this.cosmeticIndex.getSerializedSize() +
+      this.unhideIndex.getSerializedSize()
+    );
+  }
+
+  public getHTMLFilters(
+    request: Request,
+    isFilterExcluded?: (filter: NetworkFilter | CosmeticFilter) => boolean,
+  ): {
+    networkFilters: NetworkFilter[];
+    cosmeticFilters: CosmeticFilter[];
+    exceptions: NetworkFilter[];
+    unhides: CosmeticFilter[];
+  } {
+    const networkFilters: NetworkFilter[] = [];
+    const cosmeticFilters: CosmeticFilter[] = [];
+    const exceptions: NetworkFilter[] = [];
+    const unhides: CosmeticFilter[] = [];
+
+    this.networkIndex.iterMatchingFilters(request.getTokens(), (filter: NetworkFilter) => {
+      if (filter.match(request) && !isFilterExcluded?.(filter)) {
+        networkFilters.push(filter);
+      }
+      return true;
+    });
+
+    this.exceptionsIndex.iterMatchingFilters(request.getTokens(), (filter: NetworkFilter) => {
+      if (filter.match(request) && !isFilterExcluded?.(filter)) {
```

Comment:
Is it cheaper to run `isFilterExcuded` or `filter.match`?

Reply:
interesting question, we may want to benchmark it. Preprocessor exclusions were introduced recently as an optional feature, so were added after the `match`. I would not change that in this PR as this is pattern used in all buckets.